### PR TITLE
Refresh exceptions best practices

### LINKED ISSUE #40238: Improve freshness for top 10 articles
### Describe the issue or suggestion

Research the top 10 articles I own that have high page views and low freshness, and update the articles.

https://review.learn.microsoft.com/en-us/help/contribute/freshness?branch=main

Articles to refresh:

- [x] .NET Framework and Windows OS versions
- [x] .NET Framework system requirements
- [x] How to: Install and uninstall Windows services
- [x] Tutorial: Create a Windows service app
- [x] Build apps against Microsoft.NETFramework.ReferenceAssemblies
- [x] .NET project SDKs
- [x] MSBuild reference for .NET Desktop SDK projects
- [x] NETSDK1022: Duplicate items were included
- [x] Best practices for exceptions
- [x] Code-style rule options

---
[Associated WorkItem - 239907](https://dev.azure.com/msft-skilling/Content/_workitems/edit/239907)

### PR DESCRIPTION
**Pull request by gewarren**
Fixes #40238.

The main article that changed is "Best practices for exceptions". Preview is [here](https://review.learn.microsoft.com/en-us/dotnet/standard/exceptions/best-practices-for-exceptions?branch=pr-en-us-40688).

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/csharp/language-reference/statements/exception-handling-statements.md](https://github.com/dotnet/docs/blob/55e28888106741b486b080aac4eadba870b32d17/docs/csharp/language-reference/statements/exception-handling-statements.md) | [Exception-handling statements - `throw`, `try-catch`, `try-finally`, and `try-catch-finally`](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/exception-handling-statements?branch=pr-en-us-40688) |
| [docs/csharp/nullable-references.md](https://github.com/dotnet/docs/blob/55e28888106741b486b080aac4eadba870b32d17/docs/csharp/nullable-references.md) | [Nullable reference types](https://review.learn.microsoft.com/en-us/dotnet/csharp/nullable-references?branch=pr-en-us-40688) |
| [docs/fundamentals/code-analysis/quality-rules/ca1065.md](https://github.com/dotnet/docs/blob/55e28888106741b486b080aac4eadba870b32d17/docs/fundamentals/code-analysis/quality-rules/ca1065.md) | [CA1065: Do not raise exceptions in unexpected locations](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1065?branch=pr-en-us-40688) |
| [docs/fundamentals/code-analysis/quality-rules/ca2200.md](https://github.com/dotnet/docs/blob/55e28888106741b486b080aac4eadba870b32d17/docs/fundamentals/code-analysis/quality-rules/ca2200.md) | [docs/fundamentals/code-analysis/quality-rules/ca2200](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2200?branch=pr-en-us-40688) |
| [docs/fundamentals/code-analysis/quality-rules/ca2208.md](https://github.com/dotnet/docs/blob/55e28888106741b486b080aac4eadba870b32d17/docs/fundamentals/code-analysis/quality-rules/ca2208.md) | [CA2208: Instantiate argument exceptions correctly](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2208?branch=pr-en-us-40688) |
| [docs/fundamentals/code-analysis/quality-rules/index.md](https://github.com/dotnet/docs/blob/55e28888106741b486b080aac4eadba870b32d17/docs/fundamentals/code-analysis/quality-rules/index.md) | [Code quality rules overview](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/index?branch=pr-en-us-40688) |
| [docs/standard/exceptions/best-practices-for-exceptions.md](https://github.com/dotnet/docs/blob/55e28888106741b486b080aac4eadba870b32d17/docs/standard/exceptions/best-practices-for-exceptions.md) | [Best practices for exceptions](https://review.learn.microsoft.com/en-us/dotnet/standard/exceptions/best-practices-for-exceptions?branch=pr-en-us-40688) |
| [docs/standard/exceptions/how-to-use-finally-blocks.md](https://github.com/dotnet/docs/blob/55e28888106741b486b080aac4eadba870b32d17/docs/standard/exceptions/how-to-use-finally-blocks.md) | [How to use finally blocks](https://review.learn.microsoft.com/en-us/dotnet/standard/exceptions/how-to-use-finally-blocks?branch=pr-en-us-40688) |

### CODE REVIEW

**Comment by gewarren** [comment]
@stephentoub If you have time to review the updated "Best practices for exceptions" article, that would be great. The preview is [here](https://review.learn.microsoft.com/en-us/dotnet/standard/exceptions/best-practices-for-exceptions?branch=pr-en-us-40688).

**Review by BillWagner — APPROVED** [review]
This LGTM @gewarren 

Let's :shipit:

File: docs/standard/exceptions/best-practices-for-exceptions.md [hunk 1]
```diff
@@ -1,129 +1,74 @@
...
 
 Clean up resources that are allocated with either `using` statements or `finally` blocks. Prefer `using` statements to automatically clean up resources when exceptions are thrown. Use `finally` blocks to clean up resources that don't implement <xref:System.IDisposable>. Code in a `finally` clause is almost always executed even when exceptions are thrown.
 
-## Handle common conditions without throwing exceptions
+### Handle common conditions to avoid exceptions
 
-For conditions that are likely to occur but might trigger an exception, consider handling them in a way that will avoid the exception. For example, if you try to close a connection that's already closed, you'll get an `InvalidOperationException`. You can avoid that by using an `if` statement to check the connection state before trying to close it.
+For conditions that are likely to occur but might trigger an exception, consider handling them in a way that avoids the exception. For example, if you try to close a connection that's already closed, you'll get an `InvalidOperationException`. You can avoid that by using an `if` statement to check the connection state before trying to close it.
```
Comment (stephentoub) [inline docs/standard/exceptions/best-practices-for-exceptions.md:37]: > You can avoid that

Most of the time. In many of these situations there can be race conditions where such an up-front check will eliminate the exception 99% of the time, but if between the check and the operation the guarded condition changed, you could still incur an exception.

File: docs/standard/exceptions/best-practices-for-exceptions.md [hunk 2]
```diff
@@ -249,6 +194,122 @@ File name: 'C:\temp\file.txt'
...
+
+Write clear sentences and include ending punctuation. Each sentence in the string assigned to the <xref:System.Exception.Message?displayProperty=nameWithType> property should end in a period. For example, "The log table has overflowed." uses correct grammar and punctuation.
+
+### Place throw statements well
+
+Place throw statements where the stack trace will be helpful. The stack trace begins at the statement where the exception is thrown and ends at the `catch` statement that catches the exception.
+
+### Rethrow to preserve stack details
+
+Once an exception is thrown, part of the information it carries is the stack trace. The stack trace is a list of the method call hierarchy that starts with the method that throws the exception and ends with the method that catches the exception. If an exception is rethrown by specifying the exception in the `throw` statement, the stack trace is restarted at the current method and the list of method calls between the original method that threw the exception and the current method is lost. To keep the original stack trace information with the exception, use the `throw` statement without specifying the exception.
```
Comment (stephentoub) [inline docs/standard/exceptions/best-practices-for-exceptions.md:268]: > To keep the original stack trace information with the exception, use the `throw` statement without specifying the exception.

This is only relevant if you're inside of the catch block that's caught that exception instance.

**Review by stephentoub — APPROVED** [review]

**Comment by gewarren** [comment]
Thanks @stephentoub!

**Pull request merged**